frontend/src/components/auth/FirebaseAction/FirebaseAction.tsx

```diff
@@ -0,0 +1,98 @@
+import { useMutation, useQuery } from "@apollo/client";
+import React, { useEffect, useState } from "react";
+import {
+  VERIFY_EMAIL,
+  VERIFY_PASSWORD_RESET,
+} from "../../../APIClients/mutations/AuthMutations";
+import { GET_USER_BY_EMAIL } from "../../../APIClients/queries/UserQueries";
+import { Role } from "../../../types/AuthTypes";
+import LoadingState from "../../common/LoadingState";
+import NotFound from "../../pages/NotFound";
+import SetNewPassword from "../ResetPassword/SetNewPassword";
+import AdminSignupConfirmation from "../SignupConfirmation/AdminSignupConfirmation";
+import TeacherSignupConfirmation from "../SignupConfirmation/TeacherSignupConfirmation";
+import FirebaseActionError from "./FirebaseActionError";
+
+const FirebaseAction = (): React.ReactElement => {
+  const urlParams = new URLSearchParams(window.location.search);
+  const mode = urlParams.get("mode");
+  const oobCode: string = urlParams.get("oobCode") ?? "";
+
+  const [emailVerified, setEmailVerified] = useState(false);
+  const [passwordResetVerified, setPasswordResetVerified] = useState(false);
+  const [email, setEmail] = useState("");
+  const [loading, setLoading] = useState(true);
+  const [notFound, setNotFound] = useState(false);
+  const [role, setRole] = React.useState<Role | null>(null);
+
+  useQuery(GET_USER_BY_EMAIL, {
+    variables: { email },
+    onCompleted: (data) => {
+      setRole(data.userByEmail.role);
+    },
+    skip: !!role || !email,
+  });
+
+  const [verifyEmail] = useMutation<{ verifyEmail: string }>(VERIFY_EMAIL, {
+    onCompleted(data: { verifyEmail: string }) {
+      if (data.verifyEmail !== "") {
+        setEmail(data.verifyEmail);
+        setEmailVerified(true);
+        setLoading(false);
+      }
+    },
+  });
+  const [verifyPasswordReset] = useMutation<{ verifyPasswordReset: string }>(
+    VERIFY_PASSWORD_RESET,
+    {
+      onCompleted(data: { verifyPasswordReset: string }) {
+        if (data.verifyPasswordReset !== "") {
+          setEmail(data.verifyPasswordReset);
+          setPasswordResetVerified(true);
+          setLoading(false);
+        }
+      },
+    },
+  );
+
+  useEffect(() => {
+    const handleVerifyEmail = async () => {
+      await verifyEmail({ variables: { oobCode } });
+    };
+    const handleResetPassword = async () => {
+      await verifyPasswordReset({ variables: { oobCode } });
+    };
+
+    switch (mode) {
```

Comment:
why do we handle `TeacherSignupConfirmation`, `AdminSignupConfirmation` and `PasswordReset` in the same component? since it seems like they have distinct logic and UIs, does it make sense to have users directed to different pages for different flows and calling `handleVerifyEmail` and `handleResetPassword`, etc. in those respective flows?